Sming/SmingCore/Network/WebsocketClient.cpp

```diff
@@ -0,0 +1,513 @@
+ * Parameters: msg - Message int8 array 
+ *             length - length of msg  array
+ */
+void WebsocketClient::sendBinary(uint8_t* msg, uint16_t length)
+{
+	uint8_t buffer[500];
+	uint8_t i = 0;
+	buffer[i++] = 0x80 | 0x02; //FINN = 1 and Sending Binary Data
+
+	if (length <= 125)
+	{
+		buffer[i++] = 0x80 | length; //Setting MASK bit and length
+	}
+	else
+	{
+		buffer[i++] = 0xFE; // Mask bit + 126
+		buffer[i++] = (uint8_t) (length >> 8);
+		buffer[i++] = (uint8_t) (length & 0xFF);
+	}
+	//64bit outgoing messenges not supported
+
+	byte mask[4];
+	for (uint8_t j = 0; j < 4; j++)
+	{
+		mask[j] = random(0, 255);
+		buffer[i++] = mask[j];
+	}
+	for (uint16_t j = 0; j < length; j++)
+	{
+		buffer[i++] = (msg[j] ^ mask[j % 4]);
+	}
+	send((char*) buffer, i, false); // Need support from TcpClient to send binary
+}
+
+/* Function Name: onReceive
+ * Description: It process message sent byWebsocket Server
+ *              Max length of message permitted is 0xffff (65535) bytes
+ * Parameters: buf - pbuf array 
+ */
+err_t WebsocketClient::onReceive(pbuf* buf)
+{
+	if (buf == NULL)
+	{
+		// Disconnected, close it
+		TcpClient::onReceive(buf);
+	}
+	else
+	{
+		uint16_t size = buf->tot_len;
+		char* data = new char[size + 1];
+		pbuf_copy_partial(buf, data, size, 0);
+		data[size] = '\0';
+
+		//  debugf("%s", data); //print received buffer
+		switch (Mode)
+		{
+		case ws_Connecting:
+			if (verifyKey(data, size) == true)
+			{
+				Mode = ws_Connected;
+				this->connectedcallback(Mode);
+				//   debugf("Key Verified. Websocket Handshake completed");
+				sendPing();
+			}
+			else
+			{
+				Mode = ws_Disconnected; // Handshake was not proper.
+				this->connectedcallback(Mode);
+			}
+			break;
+
+		case ws_Connected:
+			// Parsing received Websocket packet
+			uint8_t op = data[0] & 0b00001111; // Extracting Opcode
+			uint8_t fin = data[0] & 0b10000000; // Extracting Fin Bit (Single Frame)
+			// debugf("Opcode : 0x%x",op);
+			// debugf("Fin : 0x%x",fin);
+			if (op == 0x00 || op == 0x01 || op == 0x02) //Data
+			{
+				if (fin > 0)
+				{
+					//  debugf("Single frame message");
+					char masked = data[1] & 0b10000000; // extracting Mask bit
+					uint16_t len = data[1] & 0b01111111; // length of data
+					uint16_t cnt = 2;
+					if (len == 126)
+					{
+						//next 2 bytes are length
+						len = data[cnt++];
+						len << 8;
+						len = len | data[cnt++];
+					}
+					if (len == 127)
+					{
+						//next 8 bytes are length
+						debugf("64bit messenges not supported"); // Too big for Esp8266 to handle
+						return -1;
+					}
+					// debugf("Message is %d chars long",len);
+
+					//Generally server replies are not masked, but RFC does not forbid it
+					uint8_t mask[4];
+					if (masked > 0)
+					{
+						mask[0] = data[cnt++];
+						mask[1] = data[cnt++];
+						mask[2] = data[cnt++];
+						mask[3] = data[cnt++];
+					}
+
+					uint8_t rxdata[len + 1]; //Max 16bit length message, so 65kbyte ...
```

Comment:
should be carefull here 65k on stack is huge will kill esp instantly, very suited for overflow attacks.
I would allocate on heap a max size and the application must be written in such a way to never send more than x bytes